### PR TITLE
add profile protocol

### PR DESCRIPTION
**Pull request by shamilovtim**
Feature:
add protocol for profile

⛔️ Blocker: we have no way of doing anything with this yet.

### CODE REVIEW

**Comment by shamilovtim** [comment]
Closes #46

**Comment by shamilovtim** [comment]
According to the current specifications the DWN does not enforce protocols at all. Thus the only reuse these give is for access control rules (read/write) but not any kind of data validation rules. 

As such our options are to require Typescript types with every protocol and/or require a JSON schema with every protocol. We should discuss the pros / cons of either approach together and figure out the best way.

**Review by amika-sq — COMMENTED** [review]
Noticed that none of these objects have any required properties defined. 

Does that mean they're all optional? Or are they all then required? I'm not quite sure, and I can't find any documentation on this.

---

Another question: In our mockups, there are fields for "Profile name (visible only to you)" and "Public display name (visible to others)". Should those be added here? Currently, all of this information is available to everyone who asks for it, so there could be some visibility issues here allowing other people to read information that the user might not think is public.

File: src/features/protocol/profile-protocol.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  socials?: { appName: string; url: string }[];
+};
+
+export type NameProtocolType = {
+  first?: string;
+  last?: string;
+  middle?: string;
+  nickname?: string;
+  maiden?: string;
+};
```
[the remark this answers is not in the document]
Comment (shamilovtim) [reply src/features/protocol/profile-protocol.ts:16]: I actually have these here intentionally to illustrate a point. The JSON schema doesn't do anything. The question is should we adopt Typescript, adopt JSON schema, or both.

File: src/features/protocol/about.schema.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
+  "type": "object",
+  "description": "Appropriate for an About Me page in a social media app",
+  "properties": {
+    "sex": { "type": "string", "enum": ["male", "female"] },
```
[the remark this answers is not in the document]
Comment (shamilovtim) [reply src/features/protocol/about.schema.json:5]: This is "sex" not "gender." But I can change this to "gender."

**Comment by shamilovtim** [comment]
> Noticed that none of these objects have any required properties defined.
> 
> Does that mean they're all optional? Or are they all then required? I'm not quite sure, and I can't find any documentation on this.

Neither technically. They are all optional because nothing actually enforces the JSON schema. It's entirely self enforced by each client who implements their own business logic. The protocol only controls access control rules but does not validate the data. 

Each client must implement their own data validation. Or, alternatively, we must create a data validation layer which clients will implement, along with protocols (which are NOT data validation). So basically we can create a new abstraction called "protocols+validation" which is protocols plus data validation. And we should decide what we want that validation written in. I think Typescript is what I'm leaning towards. The JSON schema seems rather useless because frontends don't actually use it. They use stuff like Yup or Zod.


> 
> Another question: In our mockups, there are fields for "Profile name (visible only to you)" and "Public display name (visible to others)". Should those be added here? Currently, all of this information is available to everyone who asks for it, so there could be some visibility issues here allowing other people to read information that the user might not think is public.

Good call out. The issue I think we're going to bump into is that it's impossible to predict what people want to be private or what they don't. Thus defaulting to public here made sense at the time but its a conversation to be had.

**Comment by amika-sq** [comment]
After my chat with @csuwildcat today, I'm thinking that maybe we should keep the profile object as small as possible for the time being. 

If the profile has definitions for objects such as school, work, phone, etc, we'll need to build out an editing experience for the user to actually fill out those values. Currently, that's not in scope for the initial version of the wallet. 

From the mockups, I see a couple of pieces of information for a profile that we'll need in order to build out the screens:
* Profile Name
* Public Display Name
* Icon

Wondering what our thoughts are on limiting to just these fields that we'll need for the time being? Open to thoughts there though!

File: src/features/protocol/profile-protocol.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+export const profileProtocol = {
+  profile: {
+    definition: {
+      protocol: "./",
```
Comment (csuwildcat) [inline src/features/protocol/profile-protocol.ts:21]: The 'protocols' field is required to be a URI

Comment (shamilovtim) [reply src/features/protocol/profile-protocol.ts]: Fixed

**Comment by shamilovtim** [comment]
> How are we feeling about this protocol @shamilovtim @csuwildcat ? Is there additional work to be done? We're getting close to a point on mobile where we'll need associated profile information with an identity, so I wanted to make sure this PR keeps moving forwards 😄
> 
> As of my knowledge, the identity agent mobile application needs to store/display two things related to their profiles (see attached mockup):
> 
> 1. "Profile Label" - the label the user gives their profile. This is already a part of the `ManagedIdentity` object ([link to interface](https://github.com/TBD54566975/web5-js/blob/main/packages/agent/src/identity-manager.ts#L41))
> 2. "Public Display Name" - This does not exist on a `ManagedIdentity`, and will need to be a part of the profile protocol.
> 
> Where could we slot in the "Public Display Name" for the profile? Should that be on a different protocol (maybe something like a `ManagedIdentityMetadata` protocol?)
> 
> Screenshot 2023-09-07 at 7 14 35 AM

I'm good with deleting everything other than the 1-2 fields

**Comment by csuwildcat** [comment]
I'll take a look at this today. Not sure what the last comment was suggesting in full, but will respond with a review soon.

**Comment by shamilovtim** [comment]
@amika-sq it's weird and in flux but it should be ok to merge I think

File: src/features/dwn/profile-protocol/profile.schema.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "title": "Profile",
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "type": "object",
+  "properties": {
+    "displayName": { "type": "string" }
+  }
```
Comment (amika-sq) [inline src/features/dwn/profile-protocol/profile.schema.json:7]: I'm wondering, should we add the `did` to this as well?

I'm not really sure how these records would be stored in relation to @frankhinek's work around managed identities.

* If these records are written against the master DID, there will be multiple Profile records that we'd somehow need to correlate to a ManagedIdentity. We'd need a way to answer the question: "what managedIdentity is this profile for?" A field representing which DID the profile is for could help solve this issue. 

* If each record is written to the DWN against the individual ManagedIdentity DID, we'd need to request the profile from the DWN via the specific DID anyways. A `did` field wouldn't really be useful in this case, as we needed to make a specific request for the Profile for that specific DID anyways.

File: src/types/utils.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export type SafeOmit<T, K extends keyof T> = Omit<T, K>;
```
Comment (amika-sq) [inline src/types/utils.ts:1]: Nice, love this!

**Review by amika-sq — APPROVED** [review]
Question around how the records are meant to be written to the DWN in relation to ManagedIdentities. But overall lgtm!

File: src/features/dwn/profile-protocol/profile.schema.json (hunk 4 above)
Comment (shamilovtim) [reply src/features/dwn/profile-protocol/profile.schema.json:7]: > 

Great call out. I can't quite tell one way or another. I wonder if @frankhinek would know.

**Pull request merged**